test/e2e_federation/util.go

```diff
@@ -150,28 +139,82 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
 	return clientset.Services(namespace).Create(service)
 }
 
+func createService(clientset *fedclientset.Clientset, namespace, name string) (*v1.Service, error) {
+	if clientset == nil || len(namespace) == 0 {
+		return nil, fmt.Errorf("Internal error: invalid parameters passed to createService: clientset: %v, namespace: %v", clientset, namespace)
+	}
+	By(fmt.Sprintf("Creating federated service %q in namespace %q", name, namespace))
+
+	// Tests can be run in parallel, so we need a different nodePort for
+	// each test.
+	// we add in a array all the "available" ports
+	availablePorts := make([]int32, FederatedSvcNodePortLast-FederatedSvcNodePortFirst)
+	for i := range availablePorts {
+		availablePorts[i] = int32(FederatedSvcNodePortFirst + i)
+	}
+
+	var err error
+	// until the availablePort list is not empty, lets try to create the service
+	for len(availablePorts) > 0 {
+		// select the Id of an available port
+		i := rand.Intn(len(availablePorts))
+
+		service, err := createServiceWithNodePort(clientset, namespace, name, availablePorts[i])
+		if err == nil {
+			return service, nil
+		}
+
+		// check if service have been created in some clusters.
+		// if it's the case, delete them.
+		if found, _ := hasService(clientset, namespace, name); found {
+			if err = deleteService(clientset, namespace, name, nil); err != nil {
+				framework.ExpectNoError(err, "Deleting service %q after a partial createService() error", service.Name)
+				return nil, err
+			}
+		}
+
+		// creation failed, lets try with another port
+		// first remove from the availablePorts the port with which the creation failed
+		availablePorts = append(availablePorts[:i], availablePorts[i+1:]...)
+	}
+
+	return nil, err
+}
+
 func createServiceOrFail(clientset *fedclientset.Clientset, namespace, name string) *v1.Service {
 	service, err := createService(clientset, namespace, name)
 	framework.ExpectNoError(err, "Creating service %q in namespace %q", service.Name, namespace)
 	By(fmt.Sprintf("Successfully created federated service %q in namespace %q", name, namespace))
 	return service
 }
 
-func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, serviceName string, orphanDependents *bool) {
-	if clientset == nil || len(namespace) == 0 || len(serviceName) == 0 {
-		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteServiceOrFail: clientset: %v, namespace: %v, service: %v", clientset, namespace, serviceName))
+func hasService(clientset *fedclientset.Clientset, namespace string, serviceName string) (bool, error) {
+	_, err := clientset.Core().Services(namespace).Get(serviceName, metav1.GetOptions{})
+	if err != nil && errors.IsNotFound(err) {
+		return false, err
 	}
-	framework.Logf("Deleting service %q in namespace %v", serviceName, namespace)
+	return true, err
```

Comment:
we are returning (true, err) in case of any other errors than `IsNotFound`, which is incorrect. So please put one additional condition to check other errors.